File: segmenter.py
import cv2
import numpy as np
from scipy.stats import mode
import progressbar
import time

# this function is needed for the createTrackbar step downstream
def nothing(x):
    pass

# read the experimental image

def run(filename, num_chars, orig_filename):
    img_orig = cv2.imread(filename,1)
    img = cv2.cvtColor(img_orig, cv2.COLOR_BGR2GRAY) 

    
    dilatation_size = 1
    element = cv2.getStructuringElement(cv2.MORPH_RECT, (2*dilatation_size + 1, 2*dilatation_size+1), (dilatation_size, dilatation_size))
    im = cv2.dilate(img, element,borderType=cv2.BORDER_REFLECT)

    mask = np.zeros((np.array(im.shape)+2), np.uint8)

    cv2.floodFill(im, mask, (0,0), (255))
    im_edged = cv2.Canny(im, 30, 200) 
    contours, hierarchy = cv2.findContours(im_edged, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_NONE) 
    cv2.drawContours(im, contours, -1, (255, 255, 255), 3) 

    (r,c) = im.shape;
    for i in range(r):
        for j in range(c):
            if img_orig[i][j][0] == img_orig[0][0][0] and img_orig[i][j][1] == img_orig[0][0][1] and img_orig[i][j][2] == img_orig[0][0][2] :
                im[i][j] = 255


    # SPLIT
    character = False
    hist = []
    for j in range(c):
        ctr = 0
        filled = False
        for i in range(r):
            if(im[i][j] < 255):
                ctr = ctr + 1
                if ctr > 9:
                    filled = True
            else:
                ctr = 0
        if(filled):
            if not character:
                hist.append(j)
                character = True
        else:
            if character:
                hist.append(j)
                character = False

    size = len(hist)//2;

    #CHECK IF SPLITTTING FAILS

    components = 0
    for j in range(size):
        if (hist[2*j+1]-hist[2*j] > 12):
            components = components + 1
            newmtr = im[:,hist[2*j]:hist[2*j+1]+1]
            processChar(newmtr.copy(), j)
    if components != num_chars:
        print(components, num_chars, filename)
        

    # THIS PART IS ONLY FOR GENERATING TESTCASES
    else:
        components = 0
        for j in range(size):
            if (hist[2*j+1]-hist[2*j] > 12):
                newmtr = im[:,hist[2*j]:hist[2*j+1]+1]
                processed = processChar(newmtr.copy(), j)
                char_freq[orig_filename[components]] = char_freq[orig_filename[components]] + 1
                cv2.imwrite(str(orig_filename[components]) + '/' + str(char_freq[orig_filename[components]]) + '.png', processed)
                components = components + 1


# cv2.destroyAllWindows()


def processChar(img, index):
    color = mode(img[img != 255])[0][0]
    img_filter1 = np.where(img != color, 255, img)
    img_filter2 = np.where(img_filter1 == color, 0, img_filter1)
    square_img = make_square_np(img_filter2)
    # cv2.imshow(str(index), square_img)
    # cv2.waitKey(0)
    return square_img


def make_square_np(img):
    s = max(img.shape[0:2])
    f = np.ones((s,s),np.uint8) * 255
    ax,ay = (s - img.shape[1])//2,(s - img.shape[0])//2
    f[ay:img.shape[0]+ay,ax:ax+img.shape[1]] = img
    return f
    



# run('test/image1.png')

char_freq = {
    'Q':0,
    'W':0,
    'E':0,
    'R':0,
    'T':0,
    'Y':0,
    'U':0,
    'I':0,
    'O':0,
    'P':0,
    'A':0,
    'S':0,
    'D':0,
    'F':0,
    'G':0,
    'H':0,
    'J':0,
    'K':0,
    'L':0,
    'Z':0,
    'X':0,
    'C':0,
    'V':0,
    'B':0,
    'N':0,
    'M':0
}

ctr = 0
arr = input().split()
n = len(arr)

# for name in arr:
#     run('../train/'+ name, len(name) - 4, name)

for i in progressbar.progressbar(range(n)):
    name = arr[i]
    run('../train/'+ name, len(name) - 4, name)

File: Segmenter.cpp
#include "Segmenter.h"
#define IMAGE_SIZE 30

Segmenter::Segmenter() {}

std::vector<std::vector<float>> Segmenter::segment(std::vector<std::string> filenames) {
    std::vector<std::vector<float>> return_to_python;
    int components = 0;
    for(std::string filename : filenames) {
        cv::Mat original_image = cv::imread(filename, 1);
        cv::Mat image;
        cv::cvtColor(original_image, image, CV_BGR2GRAY);
        int rows = image.rows;
        int cols = image.cols;

        if (image.empty()) {
            std::cout << "Empty Image" << std::endl;
        }

        int dilation_size = 1;
        cv::Mat element = cv::getStructuringElement(cv::MORPH_RECT,
                        cv::Size( 2*dilation_size + 1, 2*dilation_size+1 ),
                        cv::Point( dilation_size, dilation_size ));
        cv::Mat dilated_image(rows, cols, CV_8U);
        cv::dilate(image, dilated_image, element, cv::Point(-1,-1), 1, cv::BORDER_REFLECT);

        cv::Mat floodfill_mask = cv::Mat::zeros(rows + 2, cols + 2, CV_8U);
        
        cv::floodFill(dilated_image, floodfill_mask, cv::Point(0, 0), 255);
        cv::Mat dilated_canny_image(rows, cols, CV_8U);
        cv::Canny(dilated_image, dilated_canny_image, 30, 200);

        std::vector<std::vector<cv::Point>> contours;
        std::vector<cv::Vec4i> hierarchy;
        cv::findContours(dilated_canny_image, contours, hierarchy, CV_RETR_EXTERNAL, CV_CHAIN_APPROX_NONE);
        cv::drawContours(dilated_image, contours, -1, cv::Scalar(255, 255, 255), 3);

        //remove background
        cv::Vec3b top_left = original_image.at<cv::Vec3b>(0, 0);
        for(int i = 0; i < rows; i++) {
            uchar* row = dilated_image.ptr<uchar>(i);
            cv::Vec3b* original_row = original_image.ptr<cv::Vec3b>(i);
            for(int j = 0; j < cols; j++) {
                if (top_left[0] == original_row[j][0] && top_left[1] == original_row[j][1] && top_left[2] == original_row[j][2]) {
                    row[j] = 255u;
                }
            }
        }

        bool character = false;
        std::vector<float> hist;
        for(int j = 0; j < cols; j++) {
            int ctr = 0;
            bool filled = false;
            for(int i = 0; i < rows; i++) {
                if (dilated_image.at<uchar>(i, j) < 255u) {
                    ctr++;
                    if (ctr > 9) filled = true;
                } else {
                    ctr = 0;
                }
            }
            if (filled) {
                if (!character) {
                    hist.push_back(j);
                    character = true;
                }
            } else {
                if (character) {
                    hist.push_back(j);
                    character = false;
                }
            }
        }

        int size = hist.size() / 2;
        int n_chars = 0;
        for(int t = 0; t < size; t++) {
            if (hist[2*t+1]-hist[2*t] > 12) {
                int n_cols = hist[2*t+1]-hist[2*t]+1;
                cv::Mat new_image(std::max(rows, n_cols), std::max(rows, n_cols), CV_8U, cvScalar(255));
                int x_offset = (new_image.cols - n_cols) / 2;
                int y_offset = (new_image.rows - rows) / 2;
                std::vector<int> freq(256,0);

                uchar mode_color = 0;
                for(int i=0; i<rows; i++){
                    for(int j=0, j_orig=hist[2*t]; j_orig<=hist[2*t+1]; j_orig++,j++){
                        uchar color = dilated_image.at<uchar>(i, j_orig);
                        new_image.at<uchar>(i + y_offset, j + x_offset) = color;
                        freq[color]++;
                        if(color != 255u && freq[mode_color] < freq[color]){
                            mode_color = color;
                        }
                    }
                }

                for(int i=0; i<new_image.rows; i++){
                    for(int j=0; j<new_image.cols; j++){
                        if(new_image.at<uchar>(i,j) == mode_color){
                            new_image.at<uchar>(i,j) = 0u;
                        } else {
                            new_image.at<uchar>(i,j) = 255u;
                        }
                    }
                }
                
                cv::Mat resized_image;
                cv::resize(new_image, resized_image, cv::Size(IMAGE_SIZE, IMAGE_SIZE), 0, 0, CV_INTER_LINEAR);

                return_to_python.push_back({});
                for(int i=0; i<resized_image.rows; i++){
                    for(int j=0; j<resized_image.cols; j++){
                        return_to_python[components].push_back(float(resized_image.at<uchar>(i,j) / 255));
                    }
                }
                components++;
                n_chars++;
            }
        }
        this->num_chars.push_back(n_chars);
    }
    return return_to_python;
    //cv::imshow("image", dilated_image);
    //cv::waitKey(0);
}

/*
int main() {
    Segmenter mysegmenter;
    //mysegmenter.segment("train/AAA.png");

    int n;
    std::cin >> n;
    std::string s;
    for(int i=0;i<n;i++){
        std::cin >> s;
        mysegmenter.segment("train/" + s);
    }
    
    return 0;
}
*/
//link dynamic libraries
//-lopencv_core -lopencv_imgcodecs -lopencv_imgproc -lopencv_highgui